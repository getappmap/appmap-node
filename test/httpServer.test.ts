import { type IncomingMessage, request, OutgoingMessage } from "node:http";

import { fixAppmap, integrationTest, readAppmaps, spawnAppmapNode } from "./helpers";
import { ChildProcessWithoutNullStreams } from "child_process";

integrationTest("mapping Express.js requests", async () => {
  expect.assertions(1);
  const server = await spawnServer("express.js");
  await makeRequests();
  await killServer(server);
  expect(readAppmaps()).toMatchSnapshot();
});

integrationTest("mapping node:http requests", async () => {
  expect.assertions(1);
  const server = await spawnServer("vanilla.js");
  await makeRequests();
  await killServer(server);
  expect(readAppmaps()).toMatchSnapshot();
});

integrationTest("mapping Express.js requests with remote recording", async () => {
  expect.assertions(1);
  const server = await spawnServer("express.js");
  await makeRequest("/_appmap/record", "POST");
  await makeRequests();
  const appmap = JSON.parse(await makeRequest("/_appmap/record", "DELETE")) as unknown;
  expect(fixAppmap(appmap)).toMatchSnapshot();
  await killServer(server);
});

integrationTest("mapping node:http requests with remote recording", async () => {
  expect.assertions(1);
  const server = await spawnServer("vanilla.js");
  await makeRequest("/_appmap/record", "POST");
  await makeRequests();
  const appmap = JSON.parse(await makeRequest("/_appmap/record", "DELETE")) as unknown;
  expect(fixAppmap(appmap)).toMatchSnapshot();
  await killServer(server);
});

async function makeRequests() {
  await makeRequest("");
  await makeRequest("/nonexistent");
  await makeRequest("/api/foo?param1=3&param2=4");
  await makeRequest("/api/bar", "POST", (req) => {
    req.appendHeader("Content-Type", "application/json");
    req.write(
      JSON.stringify({
        key: "value",
        obj: { foo: 42, arr: [44] },
        arr: [{ foo: 43 }, { foo: 44 }],
        heterogenous: [42, "str"],
      }),
    );
  });
}

async function spawnServer(script: string) {
  const server = spawnAppmapNode(script);
  await new Promise<void>((r) =>
    server.stdout.on("data", (chunk: Buffer) => {
      if (chunk.toString().includes("listening")) r();
    }),
  );
  return server;
}

async function makeRequest(
  path: string,
  method = "GET",
  prepare?: (req: OutgoingMessage) => void,
): Promise<string> {
  const url = new URL(path, "http://127.0.0.1:27627");
  const response = new Promise<IncomingMessage>((resolve, reject) => {
    const req = request(url, { method }, resolve).once("error", reject);
    if (prepare) prepare(req);
    req.end();
  });
  const chunks: Buffer[] = [];
  for await (const chunk of await response) chunks.push(chunk as Buffer);
  return Buffer.concat(chunks).toString();
}

function killServer(server: ChildProcessWithoutNullStreams) {
  server.kill("SIGINT");
  return new Promise<void>((r) => server.on("close", () => r()));
}
