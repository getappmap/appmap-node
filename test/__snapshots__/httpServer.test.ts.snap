// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapping Express.js requests 1`] = `
{
  "./tmp/appmap/requests/<timestamp 0> -.appmap.json": {
    "classMap": [
      {
        "children": [
          {
            "children": [
              {
                "location": "./express.js:21",
                "name": "helloWorld",
                "static": true,
                "type": "function",
              },
            ],
            "name": "express",
            "type": "class",
          },
        ],
        "name": "httpServer",
        "type": "package",
      },
    ],
    "events": [
      {
        "event": "call",
        "http_server_request": {
          "headers": {
            "Host": "127.0.0.1:27627",
          },
          "path_info": "/",
          "protocol": "HTTP/1.1",
          "request_method": "GET",
        },
        "id": 1,
        "thread_id": 0,
      },
      {
        "defined_class": "express",
        "event": "call",
        "id": 2,
        "lineno": 21,
        "method_id": "helloWorld",
        "parameters": [
          {
            "class": "IncomingMessage",
            "name": "req",
            "object_id": 1,
            "value": "[IncomingMessage: GET /]",
          },
          {
            "class": "ServerResponse",
            "name": "res",
            "object_id": 2,
            "value": "[ServerResponse: 200]",
          },
          {
            "class": "Function",
            "value": "[Function: next]",
          },
        ],
        "path": "./express.js",
        "static": true,
        "thread_id": 0,
      },
      {
        "elapsed": 31.337,
        "event": "return",
        "id": 3,
        "parent_id": 2,
        "thread_id": 0,
      },
      {
        "elapsed": 31.337,
        "event": "return",
        "http_server_response": {
          "headers": {
            "Content-Length": "12",
            "Content-Type": "text/html; charset=utf-8",
            "Etag": "W/"c-Lve95gjOVATpfV8EL5X4nxwjKHE"",
            "X-Powered-By": "Express",
          },
          "status_code": 200,
        },
        "id": 4,
        "parent_id": 1,
        "thread_id": 0,
      },
    ],
    "metadata": {
      "app": "http-server-appmap-node-test",
      "client": {
        "name": "appmap-node",
        "url": "https://github.com/getappmap/appmap-node",
        "version": "test node-appmap version",
      },
      "language": {
        "engine": "Node.js",
        "name": "javascript",
        "version": "test node version",
      },
      "name": "GET / (200) — <timestamp 0>",
      "recorder": {
        "name": "requests",
        "type": "requests",
      },
    },
    "version": "1.12",
  },
  "./tmp/appmap/requests/<timestamp 1> -nonexistent.appmap.json": {
    "classMap": [],
    "events": [
      {
        "event": "call",
        "http_server_request": {
          "headers": {
            "Host": "127.0.0.1:27627",
          },
          "path_info": "/nonexistent",
          "protocol": "HTTP/1.1",
          "request_method": "GET",
        },
        "id": 1,
        "thread_id": 0,
      },
      {
        "elapsed": 31.337,
        "event": "return",
        "http_server_response": {
          "headers": {
            "Content-Length": "150",
            "Content-Security-Policy": "default-src 'none'",
            "Content-Type": "text/html; charset=utf-8",
            "X-Content-Type-Options": "nosniff",
            "X-Powered-By": "Express",
          },
          "status_code": 404,
        },
        "id": 2,
        "parent_id": 1,
        "thread_id": 0,
      },
    ],
    "metadata": {
      "app": "http-server-appmap-node-test",
      "client": {
        "name": "appmap-node",
        "url": "https://github.com/getappmap/appmap-node",
        "version": "test node-appmap version",
      },
      "language": {
        "engine": "Node.js",
        "name": "javascript",
        "version": "test node version",
      },
      "name": "GET /nonexistent (404) — <timestamp 1>",
      "recorder": {
        "name": "requests",
        "type": "requests",
      },
    },
    "version": "1.12",
  },
  "./tmp/appmap/requests/<timestamp 2> -api-foo.appmap.json": {
    "classMap": [
      {
        "children": [
          {
            "children": [
              {
                "location": "./express.js:25",
                "name": "api",
                "static": true,
                "type": "function",
              },
            ],
            "name": "express",
            "type": "class",
          },
        ],
        "name": "httpServer",
        "type": "package",
      },
    ],
    "eventUpdates": {
      "1": {
        "event": "call",
        "http_server_request": {
          "headers": {
            "Host": "127.0.0.1:27627",
          },
          "normalized_path_info": "/api/:ident",
          "path_info": "/api/foo",
          "protocol": "HTTP/1.1",
          "request_method": "GET",
        },
        "id": 1,
        "message": [
          {
            "class": "String",
            "name": "ident",
            "value": "'foo'",
          },
          {
            "class": "String",
            "name": "param1",
            "value": "'3'",
          },
          {
            "class": "String",
            "name": "param2",
            "value": "'4'",
          },
        ],
        "thread_id": 0,
      },
    },
    "events": [
      {
        "event": "call",
        "http_server_request": {
          "headers": {
            "Host": "127.0.0.1:27627",
          },
          "path_info": "/api/foo",
          "protocol": "HTTP/1.1",
          "request_method": "GET",
        },
        "id": 1,
        "message": [
          {
            "class": "String",
            "name": "param1",
            "value": "'3'",
          },
          {
            "class": "String",
            "name": "param2",
            "value": "'4'",
          },
        ],
        "thread_id": 0,
      },
      {
        "defined_class": "express",
        "event": "call",
        "id": 2,
        "lineno": 25,
        "method_id": "api",
        "parameters": [
          {
            "class": "Object",
            "name": "query",
            "object_id": 3,
            "properties": [
              {
                "class": "String",
                "name": "param1",
              },
              {
                "class": "String",
                "name": "param2",
              },
            ],
            "value": "{ param1: '3', param2: '4' }",
          },
        ],
        "path": "./express.js",
        "static": true,
        "thread_id": 0,
      },
      {
        "elapsed": 31.337,
        "event": "return",
        "id": 3,
        "parent_id": 2,
        "return_value": {
          "class": "Object",
          "object_id": 4,
          "properties": [
            {
              "class": "String",
              "name": "api",
            },
            {
              "class": "String",
              "name": "param1",
            },
            {
              "class": "String",
              "name": "param2",
            },
          ],
          "value": "{ api: 'result', param1: '3', param2: '4' }",
        },
        "thread_id": 0,
      },
      {
        "elapsed": 31.337,
        "event": "return",
        "http_server_response": {
          "headers": {
            "Content-Length": "42",
            "Content-Type": "application/json; charset=utf-8",
            "Etag": "W/"2a-YHNH/nKM8UUG4pNEEtm91sktuKc"",
            "X-Powered-By": "Express",
          },
          "status_code": 200,
        },
        "id": 4,
        "parent_id": 1,
        "thread_id": 0,
      },
    ],
    "metadata": {
      "app": "http-server-appmap-node-test",
      "client": {
        "name": "appmap-node",
        "url": "https://github.com/getappmap/appmap-node",
        "version": "test node-appmap version",
      },
      "language": {
        "engine": "Node.js",
        "name": "javascript",
        "version": "test node version",
      },
      "name": "GET /api/foo (200) — <timestamp 2>",
      "recorder": {
        "name": "requests",
        "type": "requests",
      },
    },
    "version": "1.12",
  },
  "./tmp/appmap/requests/<timestamp 3> -api-bar.appmap.json": {
    "classMap": [
      {
        "children": [
          {
            "children": [
              {
                "location": "./express.js:29",
                "name": "postApi",
                "static": true,
                "type": "function",
              },
            ],
            "name": "express",
            "type": "class",
          },
        ],
        "name": "httpServer",
        "type": "package",
      },
    ],
    "eventUpdates": {
      "1": {
        "event": "call",
        "http_server_request": {
          "headers": {
            "Content-Type": "application/json",
            "Host": "127.0.0.1:27627",
            "Transfer-Encoding": "chunked",
          },
          "normalized_path_info": "/api/:ident",
          "path_info": "/api/bar",
          "protocol": "HTTP/1.1",
          "request_method": "POST",
        },
        "id": 1,
        "message": [
          {
            "class": "String",
            "name": "ident",
            "value": "'bar'",
          },
          {
            "class": "String",
            "name": "key",
            "value": "'value'",
          },
          {
            "class": "Object",
            "name": "obj",
            "object_id": 6,
            "properties": [
              {
                "class": "Number",
                "name": "foo",
              },
              {
                "class": "Array",
                "items": {
                  "class": "Number",
                },
                "name": "arr",
              },
            ],
            "value": "{ foo: 42, arr: [ 44 ] }",
          },
          {
            "class": "Array",
            "items": {
              "class": "Object",
              "properties": [
                {
                  "class": "Number",
                  "name": "foo",
                },
              ],
            },
            "name": "arr",
            "object_id": 7,
            "size": 2,
            "value": "[ { foo: 43 }, { foo: 44 } ]",
          },
          {
            "class": "Array",
            "name": "heterogenous",
            "object_id": 8,
            "size": 2,
            "value": "[ 42, 'str' ]",
          },
        ],
        "thread_id": 0,
      },
    },
    "events": [
      {
        "event": "call",
        "http_server_request": {
          "headers": {
            "Content-Type": "application/json",
            "Host": "127.0.0.1:27627",
            "Transfer-Encoding": "chunked",
          },
          "path_info": "/api/bar",
          "protocol": "HTTP/1.1",
          "request_method": "POST",
        },
        "id": 1,
        "thread_id": 0,
      },
      {
        "defined_class": "express",
        "event": "call",
        "id": 2,
        "lineno": 29,
        "method_id": "postApi",
        "parameters": [
          {
            "class": "Object",
            "name": "body",
            "object_id": 5,
            "properties": [
              {
                "class": "String",
                "name": "key",
              },
              {
                "class": "Object",
                "name": "obj",
                "properties": [
                  {
                    "class": "Number",
                    "name": "foo",
                  },
                  {
                    "class": "Array",
                    "items": {
                      "class": "Number",
                    },
                    "name": "arr",
                  },
                ],
              },
              {
                "class": "Array",
                "items": {
                  "class": "Object",
                  "properties": [
                    {
                      "class": "Number",
                      "name": "foo",
                    },
                  ],
                },
                "name": "arr",
              },
              {
                "class": "Array",
                "name": "heterogenous",
              },
            ],
            "value": "{
  key: 'value',
  obj: { foo: 42, arr: [Array] },
  arr: [ [Object], [Object] ],
  heterogenous: [ 42, 'str' ]
}",
          },
        ],
        "path": "./express.js",
        "static": true,
        "thread_id": 0,
      },
      {
        "elapsed": 31.337,
        "event": "return",
        "id": 3,
        "parent_id": 2,
        "return_value": {
          "class": "Object",
          "object_id": 5,
          "properties": [
            {
              "class": "String",
              "name": "key",
            },
            {
              "class": "Object",
              "name": "obj",
              "properties": [
                {
                  "class": "Number",
                  "name": "foo",
                },
                {
                  "class": "Array",
                  "items": {
                    "class": "Number",
                  },
                  "name": "arr",
                },
              ],
            },
            {
              "class": "Array",
              "items": {
                "class": "Object",
                "properties": [
                  {
                    "class": "Number",
                    "name": "foo",
                  },
                ],
              },
              "name": "arr",
            },
            {
              "class": "Array",
              "name": "heterogenous",
            },
          ],
          "value": "{
  key: 'value',
  obj: { foo: 42, arr: [Array] },
  arr: [ [Object], [Object] ],
  heterogenous: [ 42, 'str' ]
}",
        },
        "thread_id": 0,
      },
      {
        "elapsed": 31.337,
        "event": "return",
        "http_server_response": {
          "headers": {
            "Content-Length": "99",
            "Content-Type": "application/json; charset=utf-8",
            "Etag": "W/"63-avsgRi3I+MK54okDiWb1V4/K3j8"",
            "X-Powered-By": "Express",
          },
          "status_code": 200,
        },
        "id": 4,
        "parent_id": 1,
        "thread_id": 0,
      },
    ],
    "metadata": {
      "app": "http-server-appmap-node-test",
      "client": {
        "name": "appmap-node",
        "url": "https://github.com/getappmap/appmap-node",
        "version": "test node-appmap version",
      },
      "language": {
        "engine": "Node.js",
        "name": "javascript",
        "version": "test node version",
      },
      "name": "POST /api/bar (200) — <timestamp 3>",
      "recorder": {
        "name": "requests",
        "type": "requests",
      },
    },
    "version": "1.12",
  },
}
`;

exports[`mapping Express.js requests with remote recording 1`] = `
{
  "classMap": [
    {
      "children": [
        {
          "children": [
            {
              "location": "./express.js:21",
              "name": "helloWorld",
              "static": true,
              "type": "function",
            },
            {
              "location": "./express.js:25",
              "name": "api",
              "static": true,
              "type": "function",
            },
            {
              "location": "./express.js:29",
              "name": "postApi",
              "static": true,
              "type": "function",
            },
          ],
          "name": "express",
          "type": "class",
        },
      ],
      "name": "httpServer",
      "type": "package",
    },
  ],
  "eventUpdates": {
    "11": {
      "event": "call",
      "http_server_request": {
        "headers": {
          "Content-Type": "application/json",
          "Host": "127.0.0.1:27627",
          "Transfer-Encoding": "chunked",
        },
        "normalized_path_info": "/api/:ident",
        "path_info": "/api/bar",
        "protocol": "HTTP/1.1",
        "request_method": "POST",
      },
      "id": 11,
      "message": [
        {
          "class": "String",
          "name": "ident",
          "value": "'bar'",
        },
        {
          "class": "String",
          "name": "key",
          "value": "'value'",
        },
        {
          "class": "Object",
          "name": "obj",
          "object_id": 6,
          "properties": [
            {
              "class": "Number",
              "name": "foo",
            },
            {
              "class": "Array",
              "items": {
                "class": "Number",
              },
              "name": "arr",
            },
          ],
          "value": "{ foo: 42, arr: [ 44 ] }",
        },
        {
          "class": "Array",
          "items": {
            "class": "Object",
            "properties": [
              {
                "class": "Number",
                "name": "foo",
              },
            ],
          },
          "name": "arr",
          "object_id": 7,
          "size": 2,
          "value": "[ { foo: 43 }, { foo: 44 } ]",
        },
        {
          "class": "Array",
          "name": "heterogenous",
          "object_id": 8,
          "size": 2,
          "value": "[ 42, 'str' ]",
        },
      ],
      "thread_id": 0,
    },
    "7": {
      "event": "call",
      "http_server_request": {
        "headers": {
          "Host": "127.0.0.1:27627",
        },
        "normalized_path_info": "/api/:ident",
        "path_info": "/api/foo",
        "protocol": "HTTP/1.1",
        "request_method": "GET",
      },
      "id": 7,
      "message": [
        {
          "class": "String",
          "name": "ident",
          "value": "'foo'",
        },
        {
          "class": "String",
          "name": "param1",
          "value": "'3'",
        },
        {
          "class": "String",
          "name": "param2",
          "value": "'4'",
        },
      ],
      "thread_id": 0,
    },
  },
  "events": [
    {
      "event": "call",
      "http_server_request": {
        "headers": {
          "Host": "127.0.0.1:27627",
        },
        "path_info": "/",
        "protocol": "HTTP/1.1",
        "request_method": "GET",
      },
      "id": 1,
      "thread_id": 0,
    },
    {
      "defined_class": "express",
      "event": "call",
      "id": 2,
      "lineno": 21,
      "method_id": "helloWorld",
      "parameters": [
        {
          "class": "IncomingMessage",
          "name": "req",
          "object_id": 1,
          "value": "[IncomingMessage: GET /]",
        },
        {
          "class": "ServerResponse",
          "name": "res",
          "object_id": 2,
          "value": "[ServerResponse: 200]",
        },
        {
          "class": "Function",
          "value": "[Function: next]",
        },
      ],
      "path": "./express.js",
      "static": true,
      "thread_id": 0,
    },
    {
      "elapsed": 31.337,
      "event": "return",
      "id": 3,
      "parent_id": 2,
      "thread_id": 0,
    },
    {
      "elapsed": 31.337,
      "event": "return",
      "http_server_response": {
        "headers": {
          "Content-Length": "12",
          "Content-Type": "text/html; charset=utf-8",
          "Etag": "W/"c-Lve95gjOVATpfV8EL5X4nxwjKHE"",
          "X-Powered-By": "Express",
        },
        "status_code": 200,
      },
      "id": 4,
      "parent_id": 1,
      "thread_id": 0,
    },
    {
      "event": "call",
      "http_server_request": {
        "headers": {
          "Host": "127.0.0.1:27627",
        },
        "path_info": "/nonexistent",
        "protocol": "HTTP/1.1",
        "request_method": "GET",
      },
      "id": 5,
      "thread_id": 0,
    },
    {
      "elapsed": 31.337,
      "event": "return",
      "http_server_response": {
        "headers": {
          "Content-Length": "150",
          "Content-Security-Policy": "default-src 'none'",
          "Content-Type": "text/html; charset=utf-8",
          "X-Content-Type-Options": "nosniff",
          "X-Powered-By": "Express",
        },
        "status_code": 404,
      },
      "id": 6,
      "parent_id": 5,
      "thread_id": 0,
    },
    {
      "event": "call",
      "http_server_request": {
        "headers": {
          "Host": "127.0.0.1:27627",
        },
        "path_info": "/api/foo",
        "protocol": "HTTP/1.1",
        "request_method": "GET",
      },
      "id": 7,
      "message": [
        {
          "class": "String",
          "name": "param1",
          "value": "'3'",
        },
        {
          "class": "String",
          "name": "param2",
          "value": "'4'",
        },
      ],
      "thread_id": 0,
    },
    {
      "defined_class": "express",
      "event": "call",
      "id": 8,
      "lineno": 25,
      "method_id": "api",
      "parameters": [
        {
          "class": "Object",
          "name": "query",
          "object_id": 3,
          "properties": [
            {
              "class": "String",
              "name": "param1",
            },
            {
              "class": "String",
              "name": "param2",
            },
          ],
          "value": "{ param1: '3', param2: '4' }",
        },
      ],
      "path": "./express.js",
      "static": true,
      "thread_id": 0,
    },
    {
      "elapsed": 31.337,
      "event": "return",
      "id": 9,
      "parent_id": 8,
      "return_value": {
        "class": "Object",
        "object_id": 4,
        "properties": [
          {
            "class": "String",
            "name": "api",
          },
          {
            "class": "String",
            "name": "param1",
          },
          {
            "class": "String",
            "name": "param2",
          },
        ],
        "value": "{ api: 'result', param1: '3', param2: '4' }",
      },
      "thread_id": 0,
    },
    {
      "elapsed": 31.337,
      "event": "return",
      "http_server_response": {
        "headers": {
          "Content-Length": "42",
          "Content-Type": "application/json; charset=utf-8",
          "Etag": "W/"2a-YHNH/nKM8UUG4pNEEtm91sktuKc"",
          "X-Powered-By": "Express",
        },
        "status_code": 200,
      },
      "id": 10,
      "parent_id": 7,
      "thread_id": 0,
    },
    {
      "event": "call",
      "http_server_request": {
        "headers": {
          "Content-Type": "application/json",
          "Host": "127.0.0.1:27627",
          "Transfer-Encoding": "chunked",
        },
        "path_info": "/api/bar",
        "protocol": "HTTP/1.1",
        "request_method": "POST",
      },
      "id": 11,
      "thread_id": 0,
    },
    {
      "defined_class": "express",
      "event": "call",
      "id": 12,
      "lineno": 29,
      "method_id": "postApi",
      "parameters": [
        {
          "class": "Object",
          "name": "body",
          "object_id": 5,
          "properties": [
            {
              "class": "String",
              "name": "key",
            },
            {
              "class": "Object",
              "name": "obj",
              "properties": [
                {
                  "class": "Number",
                  "name": "foo",
                },
                {
                  "class": "Array",
                  "items": {
                    "class": "Number",
                  },
                  "name": "arr",
                },
              ],
            },
            {
              "class": "Array",
              "items": {
                "class": "Object",
                "properties": [
                  {
                    "class": "Number",
                    "name": "foo",
                  },
                ],
              },
              "name": "arr",
            },
            {
              "class": "Array",
              "name": "heterogenous",
            },
          ],
          "value": "{
  key: 'value',
  obj: { foo: 42, arr: [Array] },
  arr: [ [Object], [Object] ],
  heterogenous: [ 42, 'str' ]
}",
        },
      ],
      "path": "./express.js",
      "static": true,
      "thread_id": 0,
    },
    {
      "elapsed": 31.337,
      "event": "return",
      "id": 13,
      "parent_id": 12,
      "return_value": {
        "class": "Object",
        "object_id": 5,
        "properties": [
          {
            "class": "String",
            "name": "key",
          },
          {
            "class": "Object",
            "name": "obj",
            "properties": [
              {
                "class": "Number",
                "name": "foo",
              },
              {
                "class": "Array",
                "items": {
                  "class": "Number",
                },
                "name": "arr",
              },
            ],
          },
          {
            "class": "Array",
            "items": {
              "class": "Object",
              "properties": [
                {
                  "class": "Number",
                  "name": "foo",
                },
              ],
            },
            "name": "arr",
          },
          {
            "class": "Array",
            "name": "heterogenous",
          },
        ],
        "value": "{
  key: 'value',
  obj: { foo: 42, arr: [Array] },
  arr: [ [Object], [Object] ],
  heterogenous: [ 42, 'str' ]
}",
      },
      "thread_id": 0,
    },
    {
      "elapsed": 31.337,
      "event": "return",
      "http_server_response": {
        "headers": {
          "Content-Length": "99",
          "Content-Type": "application/json; charset=utf-8",
          "Etag": "W/"63-avsgRi3I+MK54okDiWb1V4/K3j8"",
          "X-Powered-By": "Express",
        },
        "status_code": 200,
      },
      "id": 14,
      "parent_id": 11,
      "thread_id": 0,
    },
  ],
  "metadata": {
    "app": "http-server-appmap-node-test",
    "client": {
      "name": "appmap-node",
      "url": "https://github.com/getappmap/appmap-node",
      "version": "test node-appmap version",
    },
    "language": {
      "engine": "Node.js",
      "name": "javascript",
      "version": "test node version",
    },
    "name": "<timestamp 0>",
    "recorder": {
      "name": "remote",
      "type": "remote",
    },
  },
  "version": "1.12",
}
`;

exports[`mapping node:http requests 1`] = `
{
  "./tmp/appmap/requests/<timestamp 0> -.appmap.json": {
    "classMap": [],
    "events": [
      {
        "event": "call",
        "http_server_request": {
          "headers": {
            "Host": "127.0.0.1:27627",
          },
          "path_info": "/",
          "protocol": "HTTP/1.1",
          "request_method": "GET",
        },
        "id": 1,
        "thread_id": 0,
      },
      {
        "elapsed": 31.337,
        "event": "return",
        "http_server_response": {
          "headers": {},
          "status_code": 200,
        },
        "id": 2,
        "parent_id": 1,
        "thread_id": 0,
      },
    ],
    "metadata": {
      "app": "http-server-appmap-node-test",
      "client": {
        "name": "appmap-node",
        "url": "https://github.com/getappmap/appmap-node",
        "version": "test node-appmap version",
      },
      "language": {
        "engine": "Node.js",
        "name": "javascript",
        "version": "test node version",
      },
      "name": "GET / (200) — <timestamp 0>",
      "recorder": {
        "name": "requests",
        "type": "requests",
      },
    },
    "version": "1.12",
  },
  "./tmp/appmap/requests/<timestamp 1> -nonexistent.appmap.json": {
    "classMap": [],
    "events": [
      {
        "event": "call",
        "http_server_request": {
          "headers": {
            "Host": "127.0.0.1:27627",
          },
          "path_info": "/nonexistent",
          "protocol": "HTTP/1.1",
          "request_method": "GET",
        },
        "id": 1,
        "thread_id": 0,
      },
      {
        "elapsed": 31.337,
        "event": "return",
        "http_server_response": {
          "headers": {},
          "status_code": 200,
        },
        "id": 2,
        "parent_id": 1,
        "thread_id": 0,
      },
    ],
    "metadata": {
      "app": "http-server-appmap-node-test",
      "client": {
        "name": "appmap-node",
        "url": "https://github.com/getappmap/appmap-node",
        "version": "test node-appmap version",
      },
      "language": {
        "engine": "Node.js",
        "name": "javascript",
        "version": "test node version",
      },
      "name": "GET /nonexistent (200) — <timestamp 1>",
      "recorder": {
        "name": "requests",
        "type": "requests",
      },
    },
    "version": "1.12",
  },
  "./tmp/appmap/requests/<timestamp 2> -api-foo.appmap.json": {
    "classMap": [],
    "events": [
      {
        "event": "call",
        "http_server_request": {
          "headers": {
            "Host": "127.0.0.1:27627",
          },
          "path_info": "/api/foo",
          "protocol": "HTTP/1.1",
          "request_method": "GET",
        },
        "id": 1,
        "message": [
          {
            "class": "String",
            "name": "param1",
            "value": "'3'",
          },
          {
            "class": "String",
            "name": "param2",
            "value": "'4'",
          },
        ],
        "thread_id": 0,
      },
      {
        "elapsed": 31.337,
        "event": "return",
        "http_server_response": {
          "headers": {},
          "status_code": 200,
        },
        "id": 2,
        "parent_id": 1,
        "thread_id": 0,
      },
    ],
    "metadata": {
      "app": "http-server-appmap-node-test",
      "client": {
        "name": "appmap-node",
        "url": "https://github.com/getappmap/appmap-node",
        "version": "test node-appmap version",
      },
      "language": {
        "engine": "Node.js",
        "name": "javascript",
        "version": "test node version",
      },
      "name": "GET /api/foo (200) — <timestamp 2>",
      "recorder": {
        "name": "requests",
        "type": "requests",
      },
    },
    "version": "1.12",
  },
  "./tmp/appmap/requests/<timestamp 3> -api-bar.appmap.json": {
    "classMap": [],
    "events": [
      {
        "event": "call",
        "http_server_request": {
          "headers": {
            "Content-Type": "application/json",
            "Host": "127.0.0.1:27627",
            "Transfer-Encoding": "chunked",
          },
          "path_info": "/api/bar",
          "protocol": "HTTP/1.1",
          "request_method": "POST",
        },
        "id": 1,
        "thread_id": 0,
      },
      {
        "elapsed": 31.337,
        "event": "return",
        "http_server_response": {
          "headers": {},
          "status_code": 204,
        },
        "id": 2,
        "parent_id": 1,
        "thread_id": 0,
      },
    ],
    "metadata": {
      "app": "http-server-appmap-node-test",
      "client": {
        "name": "appmap-node",
        "url": "https://github.com/getappmap/appmap-node",
        "version": "test node-appmap version",
      },
      "language": {
        "engine": "Node.js",
        "name": "javascript",
        "version": "test node version",
      },
      "name": "POST /api/bar (204) — <timestamp 3>",
      "recorder": {
        "name": "requests",
        "type": "requests",
      },
    },
    "version": "1.12",
  },
}
`;

exports[`mapping node:http requests with remote recording 1`] = `
{
  "classMap": [],
  "events": [
    {
      "event": "call",
      "http_server_request": {
        "headers": {
          "Host": "127.0.0.1:27627",
        },
        "path_info": "/",
        "protocol": "HTTP/1.1",
        "request_method": "GET",
      },
      "id": 1,
      "thread_id": 0,
    },
    {
      "elapsed": 31.337,
      "event": "return",
      "http_server_response": {
        "headers": {},
        "status_code": 200,
      },
      "id": 2,
      "parent_id": 1,
      "thread_id": 0,
    },
    {
      "event": "call",
      "http_server_request": {
        "headers": {
          "Host": "127.0.0.1:27627",
        },
        "path_info": "/nonexistent",
        "protocol": "HTTP/1.1",
        "request_method": "GET",
      },
      "id": 3,
      "thread_id": 0,
    },
    {
      "elapsed": 31.337,
      "event": "return",
      "http_server_response": {
        "headers": {},
        "status_code": 200,
      },
      "id": 4,
      "parent_id": 3,
      "thread_id": 0,
    },
    {
      "event": "call",
      "http_server_request": {
        "headers": {
          "Host": "127.0.0.1:27627",
        },
        "path_info": "/api/foo",
        "protocol": "HTTP/1.1",
        "request_method": "GET",
      },
      "id": 5,
      "message": [
        {
          "class": "String",
          "name": "param1",
          "value": "'3'",
        },
        {
          "class": "String",
          "name": "param2",
          "value": "'4'",
        },
      ],
      "thread_id": 0,
    },
    {
      "elapsed": 31.337,
      "event": "return",
      "http_server_response": {
        "headers": {},
        "status_code": 200,
      },
      "id": 6,
      "parent_id": 5,
      "thread_id": 0,
    },
    {
      "event": "call",
      "http_server_request": {
        "headers": {
          "Content-Type": "application/json",
          "Host": "127.0.0.1:27627",
          "Transfer-Encoding": "chunked",
        },
        "path_info": "/api/bar",
        "protocol": "HTTP/1.1",
        "request_method": "POST",
      },
      "id": 7,
      "thread_id": 0,
    },
    {
      "elapsed": 31.337,
      "event": "return",
      "http_server_response": {
        "headers": {},
        "status_code": 204,
      },
      "id": 8,
      "parent_id": 7,
      "thread_id": 0,
    },
  ],
  "metadata": {
    "app": "http-server-appmap-node-test",
    "client": {
      "name": "appmap-node",
      "url": "https://github.com/getappmap/appmap-node",
      "version": "test node-appmap version",
    },
    "language": {
      "engine": "Node.js",
      "name": "javascript",
      "version": "test node version",
    },
    "name": "<timestamp 0>",
    "recorder": {
      "name": "remote",
      "type": "remote",
    },
  },
  "version": "1.12",
}
`;
